### ISSUE TITLE
Adobe Reader cannot open some pdf created with PdfPig

### ISSUE THREAD

**Issue by fviollette** [UglyToad/PdfPig#344]
Hi, 

I'm using pdfPig to cut out one big Pdf file into several pdf. 
The code below is an example of how I am using your library to create one new pdf with only the first page of my source file : 

```
            var stream = File.OpenRead(pdfPath);
            var sourceDocument = PdfDocument.Open(stream);
            var documentBuilder = new PdfDocumentBuilder();
            documentBuilder.AddPage(sourceDocument, 1);
            var newDoc = documentBuilder.Build();
            // then create fileStream from newDoc byte[]
```
The source file can be open without any error in Adobe Reader or Chrome. 
But, when I try to open the newly created with Adobe Reader, I encountered this error (It works fine with chrome btw) :

![image](https://user-images.githubusercontent.com/48249095/124279147-6f377b00-db47-11eb-9074-1af0cee7c792.png)

It works fine with many source Pdf but sometimes I'm facing this issue. It depends on the source file 😞 . 
Unfortunatly, due to the sensitive content of these pdf I cannot attached it. 

Have you ever facing such an issue ? 
Let me know if you need more information. 

Thanks in advance, 

Florent

**Comment by cremor** [comment]
I have the same problem with some PDFs that I create with `documentBuilder.AddPage(sourceDocument, pageNumber)`. For me Acrobat Reader shows the error "There was a problem reading this document (14).". 

But I noticed that the problem only happens if I use `documentBuilder.AddPage(sourceDocument, pageNumber)`. When I use `outputDocument.AddPage(sourcePage.Width, sourcePage.Height).CopyFrom(sourcePage)` instead the resulting PDF works fine!

Why are there two different methods to copy a page from a source document? Or better: Why do the two methods behave differently?
@plaisted and @InusualZ seems like you two implemented those two methods?

Some more information that I found for the broken PDFs:

- For now I only get this error for source PDFs that were created by the "Canon SC1011" scanner according to their metadata. Acrobat Reader shows those files without problems before they are merged with PdfPig.
  Sadly, I also can't share the source PDFs since they containt sensitive content.
- Ctrl-clicking on the "OK" button of that error message in Acrobat Reader gives the further information "Expected a dict object."
- Although Acrobat Reader doesn't show the PDF, other programes (e.g. Firefox) can show the PDF fine.
- When I open such a broken output PDF in the "debugger-app-3.0.0-RC1.jar" from https://pdfbox.apache.org/download.html it shows the error "1-based index not found: x" where "x" is a number like "1" or "46". It also shows the following stack trace:

```
java.lang.IllegalStateException: 1-based index not found: 46
    org.apache.pdfbox.pdmodel.PDPageTree.get(PDPageTree.java:316)
    org.apache.pdfbox.pdmodel.PDPageTree.get(PDPageTree.java:297) // This line is not always present in the stack trace!
    org.apache.pdfbox.pdmodel.PDPageTree.get(PDPageTree.java:243)
    org.apache.pdfbox.debugger.ui.DocumentEntry.getPage(DocumentEntry.java:47)
    org.apache.pdfbox.debugger.ui.PDFTreeModel.getChild(PDFTreeModel.java:139)
    [various methods from here on]
```

**Comment by cremor** [comment]
Sadly `PdfPageBuilder.CopyFrom()` can not be used as a workaround because that has its own problems: #350

**Comment by plaisted** [comment]
@cremor Implementations are slightly different as `documentBuilder.AddPage(sourceDocument, pageNumber)` is supposed to copy raw pages from a PDF to another in a performant way (and content streams are not editable) where the others are more friendly to editing (and slower due to fully parsing content streams).

Can you try opening the resulting bad pdfs in PdfPig with strict mode turned on and extract data from the page? I'm not familiar with PdfBox code base so would be helpful without a test PDF.

**Comment by cremor** [comment]
> Can you try opening the resulting bad pdfs in PdfPig with strict mode turned on

PdfPig is unable to open the created (broken) PDF. Both a "normal" `PdfDocument.Open()` and one with `ParsingOptions.LenientParsingOff` throw the same exception:

```
UglyToad.PdfPig.Core.PdfDocumentFormatException
  HResult=0x80131500
  Message=Could not find dictionary associated with reference in pages kids array: 14 0.
  Source=UglyToad.PdfPig
  StackTrace:
   at UglyToad.PdfPig.Parser.CatalogFactory.ProcessPagesNode(IndirectReference reference, DictionaryToken nodeDictionary, IndirectReference parentReference, Boolean isRoot, IPdfTokenScanner pdfTokenScanner, Boolean isLenientParsing, Int32& pageNumber)
   at UglyToad.PdfPig.Parser.CatalogFactory.Create(IndirectReference rootReference, DictionaryToken dictionary, IPdfTokenScanner scanner, Boolean isLenientParsing)
   at UglyToad.PdfPig.Parser.PdfDocumentFactory.OpenDocument(IInputBytes inputBytes, ISeekableTokenScanner scanner, ILog log, Boolean isLenientParsing, IReadOnlyList`1 passwords, Boolean clipPaths)
   at UglyToad.PdfPig.Parser.PdfDocumentFactory.Open(IInputBytes inputBytes, ParsingOptions options)
   at UglyToad.PdfPig.Parser.PdfDocumentFactory.Open(String filename, ParsingOptions options)
   at UglyToad.PdfPig.PdfDocument.Open(String filePath, ParsingOptions options)
```

**Comment by plaisted** [comment]
This will be very difficult to troubleshoot without a PDF I'm able to recreate the issue in. Can you copy the object data for `14 0` in the PDF that errored? It shouldn't have any personal information and look like:
```
14 0 obj
// bunch of data
endobj
```

**Comment by plaisted** [comment]
I believe I've identified fixes for both of these, will try to get a PR open this week.

**Comment by plaisted** [comment]
@cremor Should be good, you can test #351 to see if it resolves your issue (fixes 2 bugs, think one of them was what was causing your issue). If it doesn't fix it send me the `14 0` data and I'll see if I can find the issue.

**Comment by cremor** [comment]
I can confirm that your PR fixes the bug with my PDFs. Thanks!

**Comment by fviollette** [comment]
@plaisted do you think it will also resolves my issue : 

![image](https://user-images.githubusercontent.com/48249095/128019475-33a613c0-e490-4e2b-9970-e1bff03a5bc1.png)

If not, what kind of information do you need from me  ?

**Comment by plaisted** [comment]
@fviollette I believe it should, it resolved the same error message in Acrobat from a test PDF I used. If you still have issues let me know and I can get more info.

**Comment by fviollette** [comment]
@plaisted I have made some test and your PR fixes the bug with my pdf. 
Thanks ! And I hope it will be merged soon 😃

**Comment by cremor** [comment]
@plaisted Sadly I've just found another PDF which is not fixed by your PR. (And now the PR was merged while I was typing this comment ☹️ )

The source PDF was created with the application "Wondershare PDFelement" by merging multiple PDFs (mostly scans). The PDF was then modified by adding multiple highlights and annotations.
As before I can't share the PDF since it contains sensitive content. ☹️ 

This time the source PDF produces two different error messages in Acrobat Reader:

On page 1:
- This page was created by "Wondershare PDFelement" during the merge process. It's a table of contents (called "catalog" by PDFelement).
- Error message: "There was an error processing a page. There was a problem reading this document (110)."
- Detail from Ctrl+Click: "Unrecognized object name."
- This page is shown as blank in Acrobat Reader.

On a later page:
- This page is a scan and contains some annotations (a textbox, an arrow and multiple highlights).
- Error message: "There was an error processing an annotation or link. There was a problem reading this document (110)."
- Detail from Ctrl+Click: "Unrecognized object name."
- This page is shown in Acrobat Reader, but it is missing the annotation textbox and arrow that was added to the source PDF (but the highlights are shown).


But there is a big difference this time:
The created PDF can be opened with PdfPig (even with `LenientParsingOff`). The PDFBox Debugger app can open it too.
So I assume the PDF is valid according to the standards of PdfPig and PDFBox, but not according to Acrobat Reader.
Because of this I debugged some more and found the following:

A) Calling `GetPage(1)` the **source** PDF throws the following exception. But calling `GetPage(1)` on the created PDF does not throw an exception. So PdfPig and Acrobat Reader are exactly opposite here.
```
UglyToad.PdfPig.Core.PdfDocumentFormatException
  HResult=0x80131500
  Message=Invalid type of toUnicode CMap encountered. Got: 340 0.
  Source=UglyToad.PdfPig
  StackTrace:
   at UglyToad.PdfPig.PdfFonts.Parser.Handlers.Type0FontHandler.Generate(DictionaryToken dictionary)
   at UglyToad.PdfPig.PdfFonts.FontFactory.Get(DictionaryToken dictionary)
   at UglyToad.PdfPig.Content.ResourceStore.LoadFontDictionary(DictionaryToken fontDictionary)
   at UglyToad.PdfPig.Content.ResourceStore.LoadResourceDictionary(DictionaryToken resourceDictionary)
   at UglyToad.PdfPig.Parser.PageFactory.Create(Int32 number, DictionaryToken dictionary, PageTreeMembers pageTreeMembers, Boolean clipPaths)
   at UglyToad.PdfPig.Content.Pages.GetPage(Int32 pageNumber, Boolean clipPaths)
   at UglyToad.PdfPig.PdfDocument.GetPage(Int32 pageNumber)
```

B) The later page that shows the annotation error in Acrobat Reader contains one entry **more** in `PdfDocument.Dictionary` than the same dictionary in the source PDF.
The additional dictionary entry has the key `ProcSet` and the value `{[ /PDF, /Text, /ImageB, /ImageC, /ImageI ]}`.
Also, the dictionary entry with the key `Resources` (which exists in both versions of the file) has **more** data in the created PDF. The value in the source PDF is `{38 0}` of type `IndirectReferenceToken`. The value in the created PDF is `{<ProcSet, [ /PDF, /ImageC ]>, <XObject, <Im1, 212 0>>}` of type `DictionaryToken`.

**Comment by plaisted** [comment]
@cremor It may be useful to open a separate issue for the PdfPig PdfDocumentFormatException, it may be bug (or just a malformed PDF), I'm not very familiar with the font handling side of pdfs.

Regarding the broken merged PDF, can you find the raw contents of the first page dictionary and send it? Hard to say if it'll be enough but might point me somewhere.  Realistically we need samples to be able to recreate the issue but I understand the difficulty with private pdfs.

**Comment by cremor** [comment]
I don't think that the original PDF is malformed. Every program I tried shows it without errors, just PdfPig has problems with it. Yeah, I think I'll open a separate issue for that.

How (in which format) would you like to have that raw content?
But if I send you the raw content I could also just simply send you the first page as a PDF file, right? Because it will contain the same data, wouldn't it? That might even be possible because that page only contains the table of content. I'll check with my customer.

**Comment by plaisted** [comment]
Copy / paste using notepad should be good (it shouldn't be binary), but if you can export the single page using PdfPig and still have errors that would be best. Can give you email to send so won't be on github indefintely.

**Comment by cremor** [comment]
I've now used the current master of PdfPig (at 72db3bf244953b576a0ce75356f034e18c6cd64d) to create a new PDF that only contains page 1 of the original source PDF. This single page PDF shows the same error in Acrobat Reader as the full copy shows, so hopefully this is enough for you to find the problem.

Here is the file: [Page1.pdf](https://github.com/UglyToad/PdfPig/files/6936598/Page1.pdf)

One thing that I noticed: That single page PDF is nearly 8 MB big and the source PDF (which has 77 pages) is with 8,5 MB only a little bit bigger. If I extract the first page with other tools the created PDF is less than 50 KB small!

**Comment by plaisted** [comment]
Looks like it's related to link annotations, since the first page has links to all of other pages it copies the data over when it adds the first page (hence the large PDF) but doesn't add the pages to the page tree since the pages were added themselves.

Easiest fix would be to just strip any annotations with links to other pages but would be nice to support them in some form when copying so will need to think about that.

**Comment by cremor** [comment]
Now I'm wondering if that is actually the same problem as in my fully copied PDF. Since I copy all pages there the link annotations should be fine, right?

**Comment by plaisted** [comment]
Probably still the issue, we recreate the pages when copied to another document so the links would be invalid (wrong object number) even if the pages are copied.

**Comment by cremor** [comment]
Ok, but I don't think the link annotations alone are the problem. Because I tried to merge a few simple PDFs with "Wondershare PDFelement" myself and the merged PDF could then be copied by PdfPig without causing errors in Acrobat Reader.
Maybe the referenced pages need some specific content to cause the problem?

**Comment by cremor** [comment]
I've tried again to create a new source PDF to reproduce the problem but failed ☹️ 
I even used some of the real original scan PDFs, merged them with Wondershare PDFelement, added some annotations and then copied that file with PdfPig. But the created file works fine.

Even the links work (jump to the pages) correctly! So it seems like PdfPig does already support them?

Something is different with that one PDF that causes the problem.
Is there any chance that this is somehow related to #354? Because both problems happen with the same source PDF.

**Comment by plaisted** [comment]
I'll try to get a PR open this weekend for the annotation link issue, you can test and see if it resolves your issues, sounds like the one PDF may have something else wrong.

**Comment by plaisted** [comment]
@cremor - I updated the writer to strip the link annotations on copy, can you use my fork (https://github.com/plaisted/pdfpig) and see if it resolves your issues? It fixed the Page1.pdf you supplied me so was probably the issue. If it fixes yours I'll open a PR.

**Comment by cremor** [comment]
I've tested your change. I can confirm that it removes the link annotations (both by debugging PdfPig and by looking at the created PDF with the PDFBox debugger app) but sadly it fixes neither of the two errors shown by Acrobat Reader.

If I copy just the first page from the source PDF the created PDF is now only 25 KB small, so far so good. But that one still shows the error in Acrobat Reader too. Here is the file: 
[Copy by PdfPig with fix just page 1.pdf](https://github.com/UglyToad/PdfPig/files/6952607/Copy.by.PdfPig.with.fix.just.page.1.pdf)

**Comment by plaisted** [comment]
@cremor That was a PITA but was able to track down what the error was in the PDF, not sure what's causing though. It's creating a font dictionary missing data so likely a token it couldn't parse, PdfPig should handle this better, probably need to tweak the write logic some to ensure it doesn't create a corrupt pdf.

Can you open the source pdf and:
1. find sequence `/wspe_F2` , it should be followed by something like `7 0 R`
2. find sequence `7 0 obj` where the first two numbers are the ones found in (1)
3. copy contents from (2) between `obj` and `endobj`. It should look something like:
```
8 0 obj
<</Type /Font /Subtype /Type0 /BaseFont /MicrosoftSansSerif /Encoding /Identity-H /DescendantFonts [ 7 0 R ] /ToUnicode ????? >>
endobj
```

**Comment by cremor** [comment]
I can't find `/wspe_F2` when I open the source PDF with Notepad++.
There are 4 hits for `/wspe_F1`, but none of those have a `x 0 obj` for their x value.

If I directly search for `/Font` I find those objects (excluding those which just have `/Font x 0 R`):


```
1 0 obj
<</Type/Catalog/AcroForm<</CO[]/Fields[]/DR<</Font<</Helv 387 0 R >>>>/DA(/Helv 0 Tf 0 g)>>/Pages 2 0 R /Outlines 3 0 R /Names 5 0 R /PageLabels 384 0 R >>
endobj
```

The next object is present 4 times with different numbers in `BBox`, object reference numbers for `/wspe_F1`, `/Length` and stream data.
```
594 0 obj
<</Type/XObject/Subtype/Form/BBox[ 468.07 268.28 508.84 411.145]/Resources<</Font<</wspe_F1 598 0 R >>>>/Length 309/Filter/FlateDecode>>stream
[data]
endstream
endobj
```

The next object is present 6 times with different numbers between `startxref` and `%%EOF`. Inside the object only the object reference number for `/Metadata` is different.
```
startxref
6694014
%%EOF
1 0 obj
<</Type/Catalog/AcroForm<</CO[]/Fields[]/DR<</Font<</Helv 387 0 R >>>>/DA(/Helv 0 Tf 0 g)>>/Pages 2 0 R /Outlines 3 0 R /Names 5 0 R /PageLabels 384 0 R /Metadata 1605 0 R >>
endobj
```

```
2034 0 obj
<</Type/Page/Parent 278 0 R /MediaBox[ 0 0 595.276 841.89]/Contents 2035 0 R /Resources<</ProcSet 2036 0 R /XObject<</Im3 2037 0 R /Im8 2038 0 R /Im9 2039 0 R >>/Font<</F4 2040 0 R /F5 2041 0 R /F6 2042 0 R /F7 2047 0 R >>>>/Rotate 0/Annots[ 2072 0 R  2071 0 R  2081 0 R  2080 0 R  2090 0 R  2089 0 R ]>>
endobj
```

```
2052 0 obj
<</Type/Page/Parent 278 0 R /MediaBox[ 0 0 595.276 841.89]/Contents 2053 0 R /Resources<</ProcSet 2054 0 R /XObject<</Im3 2055 0 R /Im9 2056 0 R >>/Font<</F4 2057 0 R /F5 2058 0 R /F6 2059 0 R /F7 2064 0 R >>>>/Rotate 0>>
endobj
```

```
2052 0 obj
<</Type/Page/Parent 278 0 R /MediaBox[ 0 0 595.276 841.89]/Contents 2053 0 R /Resources<</ProcSet 2054 0 R /XObject<</Im3 2055 0 R /Im9 2056 0 R >>/Font<</F4 2057 0 R /F5 2058 0 R /F6 2059 0 R /F7 2064 0 R >>>>/Rotate 0/Annots[ 2298 0 R  2297 0 R ]>>
endobj
```

**Comment by cremor** [comment]
@plaisted Do you still plan to create a PR as mentioned in https://github.com/UglyToad/PdfPig/issues/354#issuecomment-896221116? Or do you need any additional information from me?

**Comment by plaisted** [comment]
@cremor see PR #359... may help with the bad PDF (fixes some other things).

**Comment by cremor** [comment]
As I've already commented on the PR, the PR fixes my previously reported errors, thanks!

But now I get a new error message that I didn't get before. And that is not triggered by any recent code change in PdfPig because I even get it with 0.1.5-alpha002 now. I have no idea why I didn't get it before. Maybe a recent Acrobat Reader update caused it?

Anyway, the new error is:
"Cannot extract the embedded font 'MicrosoftSansSerif'. Some characters may not display or print correctly.".
There is no additional detail message box shown by Ctrl+Click.

This error is very weird. It seems like it is only shown if I view a specific page while the annotations list is populated in Acrobat Reader. This list is populated a few seconds after opening the PDF, but only if no user interaction is happening. So it can either happen on page 1 (if nothing happens after opening the PDF) or on any other page (if you start scrolling immediately after opening the PDF).
Right now I can only reproduce that error on two specific pages of my PDF. And those two pages show different font names in the error message. If some other page is visible while the annotations list is populated, no error is shown. Even if I later scroll to one of those two pages there is still no error shown. Very weird...

I'm wondering if there is anything we (you :wink:) can do about that?

**Comment by cremor** [comment]
Some additional information for the error that I described in my last comment:

I've now figured out that this specific error is only shown if I don't change the zoom factor after opening the PDF. That doesn't make any sense for me, but I can reliably reproduce it that way. Only if the PDF is zoomed so that the page width exactly fills the window that error is shown.
Maybe it is related to the fact that showing that annotations list changes the available page width and therefore also changes the zoom factor.

I think we can leave that as an Acrobat Reader bug, right?
